### ISSUE TITLE
Cosmos: review cast/convert handling

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#34982]
In general, the logic in CosmosSqlTranslatingEV.VisitUnary for ExpressionType.Convert. See e.g. test Equality_operator_int_to_long.

**Comment by ChrisJollyAU** [comment]
@roji Would this be along the lines of a superset of the issue I picked up the other day with the string concatenation (#34963)

Also, I've tried looking but can't find the test you mentioned

**Comment by roji** [comment]
> Also, I've tried looking but can't find the test you mentioned

Yes, that test is something I'm still working on in a branch. Here it is:

```c#
[ConditionalTheory]
[MemberData(nameof(IsAsyncData))]
public virtual Task Equality_operator_int_to_long(bool async)
{
	long arg = 10248;

	return AssertQuery(
		async,
		ss => ss.Set<Order>().Where(o => o.OrderID == arg));
}
```

**Comment by roji** [comment]
> Would this be along the lines of a superset of the issue I picked up the other day with the string concatenation

And yeah, it could definitely be related...

**Comment by ChrisJollyAU** [comment]
btw, had an idea based on what I have done previously

I can get this sql and it passes

```
@__arg_0='10248'

SELECT VALUE c
FROM root c
WHERE ((c["$type"] = "Order") AND (c["OrderID"] = @__arg_0))
```

Am I correct in thinking that Cosmos has some implicit conversions, as I couldn't find much in terms of an explicit cast

**Comment by ChrisJollyAU** [comment]
@roji Submitted a PR that has a lot of improvements and lights up a bunch of tests that were failing
